test/estimate-gas.test.js

```diff
@@ -0,0 +1,138 @@
+    const tokenDecimals = 18
+    const multiplier = 10 ** tokenDecimals
+    const regFee = 1 * multiplier
+    const actFee = 1 * multiplier
+    const repReward = 1 * multiplier
+    const irnReward = repReward * 0.20
+    const mfgReward = repReward - irnReward
+    const deviceId = 'apple-iphone1'
+    const deviceIdHash = web3Utils.soliditySha3({t: 'bytes32', v: web3.fromAscii(deviceId)})
+    const mfgId = 'APPLE'
+    const deviceType = 'phone'
+    const devicePublicKey = 'somepublickey'
+    const score = 'somescore'
+
+    beforeEach(async () => {
+        ctx.contracts.token = await init.getAtonomiTokenContract(ctx.actors.owner, ctx.actors.releaseAgent)
+        ctx.contracts.atonomi = await init.getAtonomiContract(ctx.actors.owner, ctx.contracts.token.address)
+        await ctx.contracts.token.transfer(ctx.actors.deviceOwner, actFee, {from: ctx.actors.owner})
+        await ctx.contracts.token.transfer(ctx.actors.mfg, regFee + actFee, {from: ctx.actors.owner})
+        await ctx.contracts.atonomi.addNetworkMember(ctx.actors.mfg, false, true, false, mfgId, {from: ctx.actors.owner})
+        await ctx.contracts.atonomi.addNetworkMember(ctx.actors.irnNode, false, false, true, '', {from: ctx.actors.owner})
+    })
+
+    describe('global costs', () => {
+        it('   ^= to Deposit Tokens', async() => {
+            var gas = await ctx.contracts.token.transfer.estimateGas(ctx.actors.alice, regFee, {from: ctx.actors.owner})
+            console.log('      ✓ ' + gas + ' wei')
+            recordsForCSV.push( { scope: 'global', task: 'deposit tokens', gas: gas } )
+        })
+    })
+    
+    describe('atonomi costs', () => {
+        it('   ^= to Add Admin', async () => {
+            var gas = await ctx.contracts.atonomi.addNetworkMember.estimateGas(ctx.actors.admin, true, false, false, '', {from: ctx.actors.owner})
+            console.log('      ✓ ' + gas + ' wei')
+            recordsForCSV.push( { scope: 'atonomi', task: 'add admin', gas: gas } )
+        })
+        it('   ^= to Add Manufacturer', async () => {
+            var gas = await ctx.contracts.atonomi.addNetworkMember.estimateGas(ctx.actors.admin, false, true, false, '', {from: ctx.actors.owner})
+            console.log('      ✓ ' + gas + ' wei')
+            recordsForCSV.push( { scope: 'atonomi', task: 'add mfgr', gas: gas } )
+        })
+        it('   ^= to Add IRN Nodes', async () => {
+            var gas = await ctx.contracts.atonomi.addNetworkMember.estimateGas(ctx.actors.admin, false, false, true, '', {from: ctx.actors.owner})
+            console.log('      ✓ ' + gas + ' wei')
+            recordsForCSV.push( { scope: 'atonomi', task: 'add irn node', gas: gas } )
+        })
+    })
+
+    describe('manufacturer costs', () => {
+        it('   ^= to Register Device', async () => {
+            await ctx.contracts.token.contract.approve(ctx.contracts.atonomi.address, regFee, { from: ctx.actors.mfg }) //results in estimate of next line = 165980
```

Comment:
Let's add a line item on calculating the approval as well

Reply:
good call. i'm adding it for only those where approval is directly part of that transaction, not for setup of prerequisites. 

for example:
> on an owner's activateDevice() test, we ```approve()```, ```register()```, ```approve()```, ```activate()```. The estimation will only include costs for last ```approve()``` and ```activate()```.

> on a mfg's ```registerAndActivateDevice()``` test, the estimation will cover all 4 function calls.